# Credit_Card_Fraud_Detection

### I have built a Credit card Fraud Detection system using Machine Learning with Python. For this project, I have used the Logistic Regression model.

In [333]:

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from statsmodels.stats.diagnostic import linear_rainbow
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [334]:
# loading the dataset to a Pandas DataFrame
credit_card_data = pd.read_csv('creditcard.csv')

### Keggle Dataset Link
https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud/code

In [335]:
# first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [336]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [337]:
#Checking the shape of data
credit_card_data.shape

(284807, 31)

In [338]:
#Check summary
credit_card_data.describe().T

count          mean           std         min           25%  \
Time    284807.0  9.481386e+04  47488.145955    0.000000  54201.500000   
V1      284807.0  1.168375e-15      1.958696  -56.407510     -0.920373   
V2      284807.0  3.416908e-16      1.651309  -72.715728     -0.598550   
V3      284807.0 -1.379537e-15      1.516255  -48.325589     -0.890365   
V4      284807.0  2.074095e-15      1.415869   -5.683171     -0.848640   
V5      284807.0  9.604066e-16      1.380247 -113.743307     -0.691597   
V6      284807.0  1.487313e-15      1.332271  -26.160506     -0.768296   
V7      284807.0 -5.556467e-16      1.237094  -43.557242     -0.554076   
V8      284807.0  1.213481e-16      1.194353  -73.216718     -0.208630   
V9      284807.0 -2.406331e-15      1.098632  -13.434066     -0.643098   
V10     284807.0  2.239053e-15      1.088850  -24.588262     -0.535426   
V11     284807.0  1.673327e-15      1.020713   -4.797473     -0.762494   
V12     284807.0 -1.247012e-15      0.999201  -18.683715     -0.405571   
V13     284807.0  8.190001e-16      0.995274   -5.791881     -0.648539   
V14     284807.0  1.207294e-15      0.958596  -19.214325     -0.425574   
V15     284807.0  4.887456e-15      0.915316   -4.498945     -0.582884   
V16     284807.0  1.437716e-15      0.876253  -14.129855     -0.468037   
V17     284807.0 -3.772171e-16      0.849337  -25.162799     -0.483748   
V18     284807.0  9.564149e-16      0.838176   -9.498746     -0.498850   
V19     284807.0  1.039917e-15      0.814041   -7.213527     -0.456299   
V20     284807.0  6.406204e-16      0.770925  -54.497720     -0.211721   
V21     284807.0  1.654067e-16      0.734524  -34.830382     -0.228395   
V22     284807.0 -3.568593e-16      0.725702  -10.933144     -0.542350   
V23     284807.0  2.578648e-16      0.624460  -44.807735     -0.161846   
V24     284807.0  4.473266e-15      0.605647   -2.836627     -0.354586   
V25     284807.0  5.340915e-16      0.521278  -10.295397     -0.317145   
V26     284807.0  1.683437e-15      0.482227   -2.604551     -0.326984   
V27     284807.0 -3.660091e-16      0.403632  -22.565679     -0.070840   
V28     284807.0 -1.227390e-16      0.330083  -15.430084     -0.052960   
Amount  284807.0  8.834962e+01    250.120109    0.000000      5.600000   
Class   284807.0  1.727486e-03      0.041527    0.000000      0.000000   

                 50%            75%            max  
Time    84692.000000  139320.500000  172792.000000  
V1          0.018109       1.315642       2.454930  
V2          0.065486       0.803724      22.057729  
V3          0.179846       1.027196       9.382558  
V4         -0.019847       0.743341      16.875344  
V5         -0.054336       0.611926      34.801666  
V6         -0.274187       0.398565      73.301626  
V7          0.040103       0.570436     120.589494  
V8          0.022358       0.327346      20.007208  
V9         -0.051429       0.597139      15.594995  
V10        -0.092917       0.453923      23.745136  
V11        -0.032757       0.739593      12.018913  
V12         0.140033       0.618238       7.848392  
V13        -0.013568       0.662505       7.126883  
V14         0.050601       0.493150      10.526766  
V15         0.048072       0.648821       8.877742  
V16         0.066413       0.523296      17.315112  
V17        -0.065676       0.399675       9.253526  
V18        -0.003636       0.500807       5.041069  
V19         0.003735       0.458949       5.591971  
V20        -0.062481       0.133041      39.420904  
V21        -0.029450       0.186377      27.202839  
V22         0.006782       0.528554      10.503090  
V23        -0.011193       0.147642      22.528412  
V24         0.040976       0.439527       4.584549  
V25         0.016594       0.350716       7.519589  
V26        -0.052139       0.240952       3.517346  
V27         0.001342       0.091045      31.612198  
V28         0.011244       0.078280      33.847808  
Amount     22.000000      77.165000   25691.160000  

In [339]:
#Dataset informations
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [340]:
# Checking the number of missing values
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [341]:
# Checking for duplicates
credit_card_data.duplicated().sum()

1081

In [342]:
credit_card_data.drop_duplicates(inplace=True)

In [343]:
credit_card_data.duplicated().sum()

0

In [344]:
# distribution of legit transactions & fraudulent transactions
credit_card_data['Class'].value_counts()

Class
0    283253
1       473
Name: count, dtype: int64

### This Dataset is highly unblanced

**0 --> Normal Transaction**

**1 --> fraudulent transaction**

In [345]:
# separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [346]:
print(legit.shape)
print(fraud.shape)

(283253, 31)
(473, 31)


In [347]:
# statistical measures of the data
legit.Amount.describe()

count    283253.000000
mean         88.413575
std         250.379023
min           0.000000
25%           5.670000
50%          22.000000
75%          77.460000
max       25691.160000
Name: Amount, dtype: float64

In [348]:
fraud.Amount.describe()

count     473.000000
mean      123.871860
std       260.211041
min         0.000000
25%         1.000000
50%         9.820000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [349]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94835.058093  0.013439 -0.009829  0.012853 -0.010440  0.006769   
1      80450.513742 -4.498280  3.405965 -6.729599  4.472591 -2.957197   

             V6        V7        V8        V9  ...       V20      V21  \
Class                                          ...                      
0      0.001251  0.010447 -0.002448  0.002613  ... -0.000489 -0.00115   
1     -1.432518 -5.175912  0.953255 -2.522124  ...  0.405043  0.46655   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000160  0.000360  0.000393 -0.000301  0.000065  0.001409  0.000418   
1      0.086639 -0.096464 -0.106643  0.040615  0.050456  0.213774  0.078270   

           Amount  
Class              
0       88.413575  
1      123.871860  

[2 rows x 30 columns]

**Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions**

**Number of Fraudulent Transactions --> 473**

In [350]:
# Let's take Normal sample = 500
legit_sample = legit.sample(n=500)

### Concatenating two DataFrames

In [351]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [352]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
184464  126253.0  2.081873  0.228547 -2.497037  1.238975  0.879639 -1.387235   
155587  105842.0  2.018967 -0.068782 -0.306748  0.462114 -0.222270 -0.253953   
361        265.0  0.073631  1.051207 -0.281223  0.853749  1.065966  1.219197   
132000   79817.0  0.990222 -1.072905  0.310701 -1.313616 -1.074117 -0.373407   
25589    33678.0  1.294396  0.212364 -0.176356  0.313638  0.176378 -0.107806   

              V7        V8        V9  ...       V21       V22       V23  \
184464  0.991255 -0.496574  0.034534  ...  0.271710  0.861351 -0.261143   
155587 -0.530240 -0.145238  2.302890  ... -0.281054 -0.424107  0.389779   
361    -1.225597 -2.262214 -0.584441  ... -1.150128  0.870673 -0.266733   
132000 -0.418202 -0.078168  1.909399  ...  0.170522  0.387955 -0.369710   
25589  -0.095762  0.063051 -0.022105  ... -0.321892 -0.998964 -0.018287   

             V24       V25       V26       V27       V28  Amount  Class  
184464 -0.420492  0.846399 -0.162548 -0.067544 -0.084200   22.50      0  
155587  0.585852 -0.392782 -0.677710  0.010023 -0.028528    3.31      0  
361    -1.048732  0.232705 -0.262463  0.187976  0.231428    1.00      0  
132000 -0.359135  0.548234  0.146179  0.014165  0.046508  180.00      0  
25589  -0.913283  0.313401  0.152700 -0.033757  0.006942    1.98      0  

[5 rows x 31 columns]

In [353]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [354]:
new_dataset['Class'].value_counts()

Class
0    500
1    473
Name: count, dtype: int64

In [355]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92297.778000  0.001748 -0.123335 -0.023114  0.047919  0.017169   
1      80450.513742 -4.498280  3.405965 -6.729599  4.472591 -2.957197   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.019760 -0.004632  0.065471 -0.071343  ... -0.044660 -0.047655   
1     -1.432518 -5.175912  0.953255 -2.522124  ...  0.405043  0.466550   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.039076 -0.009678  0.021875  0.007601 -0.001126  0.019462 -0.011719   
1      0.086639 -0.096464 -0.106643  0.040615  0.050456  0.213774  0.078270   

          Amount  
Class             
0       87.71516  
1      123.87186  

[2 rows x 30 columns]

### Splitting the data into Features & Targets

In [356]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [357]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
184464  126253.0  2.081873  0.228547 -2.497037  1.238975  0.879639 -1.387235   
155587  105842.0  2.018967 -0.068782 -0.306748  0.462114 -0.222270 -0.253953   
361        265.0  0.073631  1.051207 -0.281223  0.853749  1.065966  1.219197   
132000   79817.0  0.990222 -1.072905  0.310701 -1.313616 -1.074117 -0.373407   
25589    33678.0  1.294396  0.212364 -0.176356  0.313638  0.176378 -0.107806   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [358]:
print(Y)

184464    0
155587    0
361       0
132000    0
25589     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 973, dtype: int64


### Split the data into Training data & Testing Data

In [359]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.20, stratify=Y, random_state=2)

In [360]:
print(X.shape, X_train.shape, X_test.shape)

(973, 30) (778, 30) (195, 30)


### Model Training

In [361]:
# Logistic Regression
model = LogisticRegression()

In [362]:
# Training the Logistic Regression Model with Training Data
basemodel=model.fit(X_train, Y_train)
basemodel

LogisticRegression()

### Model Evaluation

#### Accuracy Score

In [363]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [364]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9395886889460154


In [365]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [366]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9230769230769231


* **The provided code snippet outputs the accuracy score for test data, revealing a commendable accuracy rate of approximately 93.85%.**
* **This suggests that the model performs well in making correct predictions on the test dataset.**